File: ExercismFSharp/ExercismFSharp/Bob.fs
module Bob

open System

let hey (phrase : string) =
    let isUpperCaseLetter letter =
        Char.IsLetter letter && Char.IsUpper letter

    let areAllLettersUpperCase string =
        string
        |> Seq.filter Char.IsLetter
        |> Seq.forall isUpperCaseLetter

    let hasUpperCaseLetter string =
        string
        |> Seq.exists (fun n -> isUpperCaseLetter n)

    let isQuestion phrase =
        phrase |> Seq.tryLast = Some '?'

    let isYelling phrase =
         hasUpperCaseLetter phrase && areAllLettersUpperCase phrase

    let isNothing phrase =
        (Seq.isEmpty phrase) || (Seq.forall Char.IsWhiteSpace phrase)

    match phrase with
    | _ when isYelling phrase -> "Whoa, chill out!"
    | _ when isQuestion phrase -> "Sure."
    | _ when isNothing phrase -> "Fine. Be that way!"
    | _ -> "Whatever."

File: Exercism/python/bob/bob.py
def hey(phrase):
    phrase = phrase.strip()

    if len(phrase) == 0:
        return "Fine. Be that way!"

    letters = filter(lambda x: x.isalpha(), phrase)

    if len(letters) > 0 and all(letter.isupper() for letter in letters):
        if phrase.endswith("?"):
            return "Calm down, I know what I'm doing!"
        else:
            return "Whoa, chill out!"

    if phrase.endswith("?"):
        return "Sure."

    return "Whatever."
